src/sentry/utils/outcomes.py

```diff
@@ -0,0 +1,100 @@
+from sentry.utils.pubsub import QueuedPublisherService, KafkaPublisher
+
+# valid values for outcome
+ACCEPTED = 0
+FILTERED = 1
+RATE_LIMITED = 2
+INVALID = 3
+
+outcomes = settings.KAFKA_TOPICS[settings.KAFKA_OUTCOMES]
+outcomes_publisher = None
+
+
+def track_outcome(org_id, project_id, key_id, outcome, reason=None, timestamp=None):
+    """
+    This is a central point to track org/project counters per incoming event.
+    NB: This should only ever be called once per incoming event, which means
+    it should only be called at the point we know the final outcome for the
+    event (invalid, rate_limited, accepted, discarded, etc.)
+
+    This increments all the relevant legacy RedisTSDB counters, as well as
+    sending a single metric event to Kafka which can be used to reconstruct the
+    counters with SnubaTSDB.
+    """
+    global outcomes_publisher
+    if outcomes_publisher is None:
+        outcomes_publisher = QueuedPublisherService(
+            KafkaPublisher(
+                settings.KAFKA_CLUSTERS[outcomes['cluster']]
+            )
+        )
+
+    timestamp = timestamp or to_datetime(time.time())
+    increment_list = []
+    if outcome != INVALID:
+        # This simply preserves old behavior. We never counted invalid events
+        # (too large, duplicate, CORS) toward regular `received` counts.
+        increment_list.extend([
+            (tsdb.models.project_total_received, project_id),
+            (tsdb.models.organization_total_received, org_id),
+            (tsdb.models.key_total_received, key_id),
+        ])
+
+    if outcome == FILTERED:
+        increment_list.extend([
+            (tsdb.models.project_total_blacklisted, project_id),
+            (tsdb.models.organization_total_blacklisted, org_id),
+            (tsdb.models.key_total_blacklisted, key_id),
+        ])
+    elif outcome == RATE_LIMITED:
+        increment_list.extend([
+            (tsdb.models.project_total_rejected, project_id),
+            (tsdb.models.organization_total_rejected, org_id),
+            (tsdb.models.key_total_rejected, key_id),
+        ])
+
+    if reason in FILTER_STAT_KEYS_TO_VALUES:
+        increment_list.append((FILTER_STAT_KEYS_TO_VALUES[reason], project_id))
+
+    increment_list = [(model, key) for model, key in increment_list if key is not None]
+    if increment_list:
+        tsdb.incr_multi(increment_list, timestamp=timestamp)
+
+    # Send a snuba metrics payload.
+    if random.random() <= options.get('snuba.track-outcomes-sample-rate'):
+        outcomes_publisher.publish(
+            outcomes['topic'],
+            json.dumps({
+                'timestamp': timestamp,
+                'org_id': org_id,
+                'project_id': project_id,
+                'key_id': key_id,
+                'outcome': outcome,
+                'reason': reason,
+            })
+        )
+
+    metrics.incr(
+        'events.outcomes',
+        skip_internal=True,
+        tags={
+            'outcome': outcome,
```

Comment:
@JTCunning do I have to convert this into a string again for datadog?

Reply:
That would be good. Discussed offline that using an Enum class for lookups might be better.